notCompatibleWithConfigurationCache failed to force errors into warnings

### Context
Android Gradle Plugin is adopting the new API `notCompatibleWithConfigurationCache` to let Gradle know which tasks are not compatible with config caching.

### Steps to Reproduce 

With the sample project attached, "androidDependencies" task has been specified as incompatible with config cache in app build file.

1. run `./gradlew androidDependencies`, and expect the build to finish
2. run `./gradlew androidDependencies --configuration-cache`, the build would fail unexpectedly

### Expected Behavior
<!--- Tell us what should happen -->
With `notCompatibleWithConfigurationCache`, Gradle would force config cache errors into warnings

### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
The build failed hard with errors like 

```
Configuration cache state could not be cached: field '__nestedComponents__' from type 'com.android.build.gradle.internal.tasks.DependencyReportTask': error writing value of type 'org.gradle.api.internal.provider.DefaultListProperty'
> Configuration cache state could not be cached: field 'delegate$delegate' from type 'com.android.build.api.component.impl.AndroidTestImpl': error writing value of type 'kotlin.SynchronizedLazyImpl'
   > Configuration cache state could not be cached: field 'value' from type 'kotlin.InitializedLazyImpl': error writing value of type 'com.android.build.api.component.impl.AndroidTestCreationConfigImpl'
      > Configuration cache state could not be cached: field 'globalScope' from type 'com.android.build.api.component.impl.AndroidTestCreationConfigImpl': error writing value of type 'com.android.build.gradle.internal.scope.GlobalScope'
         > Configuration cache state could not be cached: field 'extension' from type 'com.android.build.gradle.internal.scope.GlobalScope': error writing value of type 'com.android.build.gradle.internal.dsl.BaseAppModuleExtension'
            > Configuration cache state could not be cached: field 'publicExtensionImpl' from type 'com.android.build.gradle.internal.dsl.BaseAppModuleExtension': error writing value of type 'com.android.build.gradle.internal.dsl.ApplicationExtensionImpl$AgpDecorated'
               > Configuration cache state could not be cached: field 'sdkComponents$delegate' from type 'com.android.build.gradle.internal.dsl.ApplicationExtensionImpl$AgpDecorated': error writing value of type 'kotlin.SynchronizedLazyImpl'
                  > Could not create an instance of type com.android.build.gradle.internal.dsl.SdkComponentsImpl.
                     > Unable to determine constructor argument #6: missing parameter of type Provider, or no service of type Provider<Provider<List<RegularFile>>>.
```

[TaskNotCompatibleWithConfigCache.tar.gz](https://github.com/gradle/gradle/files/8093423/TaskNotCompatibleWithConfigCache.tar.gz)

New findings: 

If you use the `notCompatibleWithConfigurationCache` API with ` :app:signingReport` task(just like what I did for `androidDependencies`) and run with config caching enabled, Gradle also fails to force error into warnings. See build output below.

---------------------------------------------------------------------------------------------------------------------

Configuration cache is an incubating feature.
Calculating task graph as no configuration cache is available for tasks: signingReport

149 problems were found storing the configuration cache, 3 of which seem unique.
- Task `:app:signingReport` of type `com.android.build.gradle.internal.tasks.SigningReportTask`: cannot serialize object of type 'org.gradle.api.internal.artifacts.configurations.DefaultConfiguration', a subtype of 'org.gradle.api.artifacts.Configuration', as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.4/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:app:signingReport` of type `com.android.build.gradle.internal.tasks.SigningReportTask`: cannot serialize object of type 'org.gradle.api.internal.artifacts.configurations.DefaultConfigurationContainer', a subtype of 'org.gradle.api.artifacts.ConfigurationContainer', as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.4/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:app:signingReport` of type `com.android.build.gradle.internal.tasks.SigningReportTask`: cannot serialize object of type 'org.gradle.api.internal.project.DefaultProject', a subtype of 'org.gradle.api.Project', as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.4/userguide/configuration_cache.html#config_cache:requirements:disallowed_types

See the complete report at file:///usr/local/google/home/bingran/AndroidStudioProjects/TaskNotCompatibleWithConfigCache/build/reports/configuration-cache/3wbdeznuq9yzyrp4za8zccu7k/11gmfs5xydkv4cjdfo210s5nw/configuration-cache-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache state could not be cached: field 'components' from type 'com.android.build.gradle.internal.tasks.SigningReportTask': error writing value of type 'com.google.common.collect.RegularImmutableList'
> Configuration cache state could not be cached: field 'elements' from type 'com.google.common.collect.ImmutableList$SerializedForm': error writing value of type '[Ljava.lang.Object;'
   > Configuration cache state could not be cached: field 'androidTest' from type 'com.android.build.api.variant.impl.ApplicationVariantImpl': error writing value of type 'com.android.build.api.component.impl.AndroidTestImpl'
      > Configuration cache state could not be cached: field 'delegate$delegate' from type 'com.android.build.api.component.impl.AndroidTestImpl': error writing value of type 'kotlin.SynchronizedLazyImpl'
         > Configuration cache state could not be cached: field 'value' from type 'kotlin.InitializedLazyImpl': error writing value of type 'com.android.build.api.component.impl.AndroidTestCreationConfigImpl'
            > Configuration cache state could not be cached: field 'globalScope' from type 'com.android.build.api.component.impl.AndroidTestCreationConfigImpl': error writing value of type 'com.android.build.gradle.internal.scope.GlobalScope'
               > Configuration cache state could not be cached: field 'extension' from type 'com.android.build.gradle.internal.scope.GlobalScope': error writing value of type 'com.android.build.gradle.internal.dsl.BaseAppModuleExtension'
                  > Configuration cache state could not be cached: field 'publicExtensionImpl' from type 'com.android.build.gradle.internal.dsl.BaseAppModuleExtension': error writing value of type 'com.android.build.gradle.internal.dsl.ApplicationExtensionImpl$AgpDecorated'
                     > Configuration cache state could not be cached: field 'sdkComponents$delegate' from type 'com.android.build.gradle.internal.dsl.ApplicationExtensionImpl$AgpDecorated': error writing value of type 'kotlin.SynchronizedLazyImpl'
                        > Could not create an instance of type com.android.build.gradle.internal.dsl.SdkComponentsImpl.
                           > Unable to determine constructor argument #6: missing parameter of type Provider, or no service of type Provider<Provider<List<RegularFile>>>.

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIncompatibleTasksIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIncompatibleTasksIntegrationTest.groovy
@@ -21,6 +21,43 @@ import org.gradle.integtests.fixtures.configurationcache.ConfigurationCacheFixtu
 class ConfigurationCacheIncompatibleTasksIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
     ConfigurationCacheFixture fixture = new ConfigurationCacheFixture(this)
 
+    def "reports incompatible task serialization error and discards cache entry when task is scheduled"() {
+        given:
+        withBrokenSerializableType()
+        addTasksWithProblems('new BrokenSerializable()')
+
+        when:
+        configurationCacheRun("declared")
+
+        then:
+        result.assertTasksExecuted(":declared")
+        fixture.assertStateStoredAndDiscarded {
+            problem "Task `:declared` of type `Broken`: error writing value of type 'BrokenSerializable'"
+            problem "Task `:declared` of type `Broken`: invocation of 'Task.project' at execution time is unsupported."
+        }
+
+        when:
+        configurationCacheRun("declared")
+
+        then:
+        result.assertTasksExecuted(":declared")
+        fixture.assertStateStoredAndDiscarded {
+            problem "Task `:declared` of type `Broken`: error writing value of type 'BrokenSerializable'"
+            problem "Task `:declared` of type `Broken`: invocation of 'Task.project' at execution time is unsupported."
+        }
+    }
+
+    private withBrokenSerializableType() {
+        buildFile '''
+            import java.io.*
+            @SuppressWarnings('GrMethodMayBeStatic')
+            class BrokenSerializable implements Serializable {
+                String toString() { 'BrokenSerializable' }
+                private Object writeReplace() { throw new RuntimeException('BOOM!') }
+            }
+        '''
+    }
+
     def "reports incompatible task serialization and execution problems and discards cache entry when task is scheduled"() {
         addTasksWithProblems()
 
@@ -169,7 +206,7 @@ class ConfigurationCacheIncompatibleTasksIntegrationTest extends AbstractConfigu
 
     private addTasksWithoutProblems() {
         buildFile """
-            task declared {
+            tasks.register('declared') {
                 notCompatibleWithConfigurationCache("not really")
                 doLast {
                 }
@@ -177,11 +214,11 @@ class ConfigurationCacheIncompatibleTasksIntegrationTest extends AbstractConfigu
         """
     }
 
-    private addTasksWithProblems() {
-        buildFile("""
+    private addTasksWithProblems(String brokenFieldValue = 'project.configurations') {
+        buildFile """
             class Broken extends DefaultTask {
                 // Serialization time problem
-                private final configurations = project.configurations
+                private final brokenField = $brokenFieldValue
 
                 @TaskAction
                 void execute() {
@@ -200,6 +237,6 @@ class ConfigurationCacheIncompatibleTasksIntegrationTest extends AbstractConfigu
             tasks.register("ok") {
                 doLast { }
             }
-        """)
+        """
     }
 }

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblems.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ConfigurationCacheProblems.kt
@@ -107,6 +107,10 @@ class ConfigurationCacheProblems(
             override fun onProblem(problem: PropertyProblem) {
                 onProblem(problem, ProblemSeverity.Suppressed)
             }
+
+            override fun onError(trace: PropertyTrace, error: Exception, message: StructuredMessageBuilder) {
+                onProblem(PropertyProblem(trace, StructuredMessage.build(message), error))
+            }
         }
     }
 

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ProblemsListener.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/ProblemsListener.kt
@@ -16,6 +16,8 @@
 
 package org.gradle.configurationcache.problems
 
+import org.gradle.configurationcache.ConfigurationCacheError
+import org.gradle.configurationcache.extensions.maybeUnwrapInvocationTargetException
 import org.gradle.internal.service.scopes.EventScope
 import org.gradle.internal.service.scopes.Scopes
 
@@ -24,6 +26,13 @@ import org.gradle.internal.service.scopes.Scopes
 interface ProblemsListener {
 
     fun onProblem(problem: PropertyProblem)
+
+    fun onError(trace: PropertyTrace, error: Exception, message: StructuredMessageBuilder) {
+        throw ConfigurationCacheError(
+            "${propertyDescriptionFor(trace)}: ${StructuredMessage.build(message)}",
+            error.maybeUnwrapInvocationTargetException()
+        )
+    }
 
     fun forIncompatibleType(): ProblemsListener = this
 }

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Codec.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Codec.kt
@@ -25,6 +25,7 @@ import org.gradle.configurationcache.DefaultConfigurationCache
 import org.gradle.configurationcache.extensions.uncheckedCast
 import org.gradle.configurationcache.problems.PropertyProblem
 import org.gradle.configurationcache.problems.PropertyTrace
+import org.gradle.configurationcache.problems.StructuredMessageBuilder
 import org.gradle.configurationcache.serialization.beans.BeanStateReader
 import org.gradle.configurationcache.serialization.beans.BeanStateWriter
 import org.gradle.internal.serialize.Decoder
@@ -104,6 +105,8 @@ interface IsolateContext {
     val trace: PropertyTrace
 
     fun onProblem(problem: PropertyProblem)
+
+    fun onError(error: Exception, message: StructuredMessageBuilder)
 }
 
 

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Contexts.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Contexts.kt
@@ -25,6 +25,7 @@ import org.gradle.configurationcache.ClassLoaderScopeSpec
 import org.gradle.configurationcache.problems.ProblemsListener
 import org.gradle.configurationcache.problems.PropertyProblem
 import org.gradle.configurationcache.problems.PropertyTrace
+import org.gradle.configurationcache.problems.StructuredMessageBuilder
 import org.gradle.configurationcache.serialization.beans.BeanStateReader
 import org.gradle.configurationcache.serialization.beans.BeanStateReaderLookup
 import org.gradle.configurationcache.serialization.beans.BeanStateWriter
@@ -392,6 +393,10 @@ abstract class AbstractIsolateContext<T>(
 
     override fun onProblem(problem: PropertyProblem) {
         currentProblemsListener.onProblem(problem)
+    }
+
+    override fun onError(error: Exception, message: StructuredMessageBuilder) {
+        currentProblemsListener.onError(trace, error, message)
     }
 
     override suspend fun forIncompatibleType(action: suspend () -> Unit) {

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/beans/BeanPropertyWriter.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/beans/BeanPropertyWriter.kt
@@ -19,14 +19,10 @@ package org.gradle.configurationcache.serialization.beans
 import com.google.common.primitives.Primitives.wrap
 import org.gradle.api.internal.GeneratedSubclasses
 import org.gradle.api.internal.IConventionAware
-import org.gradle.configurationcache.ConfigurationCacheError
 import org.gradle.configurationcache.ConfigurationCacheProblemsException
-import org.gradle.configurationcache.extensions.maybeUnwrapInvocationTargetException
 import org.gradle.configurationcache.extensions.uncheckedCast
 import org.gradle.configurationcache.problems.PropertyKind
-import org.gradle.configurationcache.problems.propertyDescriptionFor
 import org.gradle.configurationcache.serialization.Codec
-import org.gradle.configurationcache.serialization.IsolateContext
 import org.gradle.configurationcache.serialization.WriteContext
 import org.gradle.configurationcache.serialization.withDebugFrame
 import java.io.IOException
@@ -103,22 +99,17 @@ suspend fun WriteContext.writeNextProperty(name: String, value: Any?, kind: Prop
         } catch (passThrough: ConfigurationCacheProblemsException) {
             throw passThrough
         } catch (error: Exception) {
-            throw ConfigurationCacheError(
-                propertyErrorMessage(value),
-                error.maybeUnwrapInvocationTargetException()
-            )
+            onError(error) {
+                when {
+                    value != null -> {
+                        text("error writing value of type ")
+                        reference(GeneratedSubclasses.unpackType(value))
+                    }
+                    else -> {
+                        text("error writing null value")
+                    }
+                }
+            }
         }
     }
 }
-
-
-private
-fun IsolateContext.propertyErrorMessage(value: Any?) =
-    "${propertyDescriptionFor(trace)}: error writing value of type '${
-    value?.let { unpackedTypeNameOf(it) } ?: "null"
-    }'"
-
-
-private
-fun unpackedTypeNameOf(value: Any) =
-    GeneratedSubclasses.unpackType(value).name

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/CachedEnvironmentStateCodec.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/CachedEnvironmentStateCodec.kt
@@ -16,11 +16,8 @@
 
 package org.gradle.configurationcache.serialization.codecs
 
-import org.gradle.configurationcache.ConfigurationCacheError
 import org.gradle.configurationcache.ConfigurationCacheProblemsException
-import org.gradle.configurationcache.extensions.maybeUnwrapInvocationTargetException
 import org.gradle.configurationcache.problems.PropertyTrace
-import org.gradle.configurationcache.problems.propertyDescriptionFor
 import org.gradle.configurationcache.serialization.Codec
 import org.gradle.configurationcache.serialization.ReadContext
 import org.gradle.configurationcache.serialization.WriteContext
@@ -33,11 +30,13 @@ import java.io.IOException
 
 internal
 object CachedEnvironmentStateCodec : Codec<EnvironmentChangeTracker.CachedEnvironmentState> {
+
     override suspend fun WriteContext.encode(value: EnvironmentChangeTracker.CachedEnvironmentState) {
         writeBoolean(value.cleared)
 
         writeCollection(value.updates) { update ->
-            withPropertyTrace(PropertyTrace.SystemProperty(update.key.toString(), update.location ?: PropertyTrace.Unknown)) {
+            val keyString = update.key.toString()
+            withPropertyTrace(PropertyTrace.SystemProperty(keyString, update.location ?: PropertyTrace.Unknown)) {
                 try {
                     write(update.key)
                     write(update.value)
@@ -46,10 +45,10 @@ object CachedEnvironmentStateCodec : Codec<EnvironmentChangeTracker.CachedEnviro
                 } catch (passThrough: ConfigurationCacheProblemsException) {
                     throw passThrough
                 } catch (error: Exception) {
-                    throw ConfigurationCacheError(
-                        propertyDescriptionFor(trace),
-                        error.maybeUnwrapInvocationTargetException()
-                    )
+                    onError(error) {
+                        text("failed to write system property ")
+                        reference(keyString)
+                    }
                 }
             }
         }

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt
@@ -226,10 +226,10 @@ suspend fun <T> T.withTaskOf(
 ) where T : IsolateContext, T : MutableIsolateContext {
     withIsolate(IsolateOwner.OwnerTask(task), codec) {
         withPropertyTrace(PropertyTrace.Task(taskType, task.path)) {
-            if (!task.isCompatibleWithConfigurationCache) {
-                forIncompatibleType(action)
-            } else {
+            if (task.isCompatibleWithConfigurationCache) {
                 action()
+            } else {
+                forIncompatibleType(action)
             }
         }
     }

--- a/subprojects/configuration-cache/src/test/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintCheckerTest.kt
+++ b/subprojects/configuration-cache/src/test/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintCheckerTest.kt
@@ -27,6 +27,7 @@ import org.gradle.api.provider.ValueSourceParameters
 import org.gradle.configurationcache.CheckedFingerprint
 import org.gradle.configurationcache.problems.PropertyProblem
 import org.gradle.configurationcache.problems.PropertyTrace
+import org.gradle.configurationcache.problems.StructuredMessageBuilder
 import org.gradle.configurationcache.serialization.Codec
 import org.gradle.configurationcache.serialization.IsolateOwner
 import org.gradle.configurationcache.serialization.CircularReferences
@@ -306,6 +307,9 @@ class ConfigurationCacheFingerprintCheckerTest {
         override fun onProblem(problem: PropertyProblem): Unit =
             undefined()
 
+        override fun onError(error: Exception, message: StructuredMessageBuilder) =
+            undefined()
+
         override fun push(codec: Codec<Any?>): Unit =
             undefined()
 
@@ -407,6 +411,9 @@ class ConfigurationCacheFingerprintCheckerTest {
             set(_) {}
 
         override fun onProblem(problem: PropertyProblem): Unit =
+            undefined()
+
+        override fun onError(error: Exception, message: StructuredMessageBuilder) =
             undefined()
 
         override fun push(codec: Codec<Any?>): Unit =

